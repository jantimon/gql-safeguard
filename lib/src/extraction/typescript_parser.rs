//! TypeScript/TSX Parser for GraphQL Template Literals
//!
//! Uses SWC AST parsing instead of regex to avoid false positives from commented GraphQL.

use anyhow::Result;
use std::fs;
use std::path::Path;
use swc_core::ecma::{
    ast::*,
    parser::{lexer::Lexer, Parser, StringInput, Syntax, TsConfig},
    visit::{Visit, VisitWith},
};

#[derive(Debug, Clone)]
pub struct GraphQLString {
    pub content: String,
    pub file_path: std::path::PathBuf,
    pub position: u32,
}

/// Extracts GraphQL from `gql` and `graphql` tagged template literals.
/// Skips templates with interpolation since they can't be statically analyzed.
pub fn extract_graphql_from_file(file_path: &Path) -> Result<Vec<GraphQLString>> {
    let source_code = fs::read_to_string(file_path)?;

    let syntax = if file_path.extension().and_then(|s| s.to_str()) == Some("tsx") {
        Syntax::Typescript(TsConfig {
            tsx: true,
            ..Default::default()
        })
    } else {
        Syntax::Typescript(TsConfig {
            tsx: false,
            ..Default::default()
        })
    };

    let lexer = Lexer::new(
        syntax,
        Default::default(),
        StringInput::new(&source_code, Default::default(), Default::default()),
        None,
    );

    let mut parser = Parser::new_from(lexer);
    let module = parser.parse_module().map_err(|e| {
        anyhow::anyhow!("TypeScript parse error in {}: {:?}", file_path.display(), e)
    })?;

    let mut visitor = GraphQLVisitor::new(file_path.to_path_buf());
    module.visit_with(&mut visitor);

    Ok(visitor.graphql_strings)
}

struct GraphQLVisitor {
    file_path: std::path::PathBuf,
    graphql_strings: Vec<GraphQLString>,
}

impl GraphQLVisitor {
    /// Creates a new visitor for the specified file path.
    ///
    /// # Arguments
    /// * `file_path` - Path to the file being processed
    fn new(file_path: std::path::PathBuf) -> Self {
        Self {
            file_path,
            graphql_strings: Vec::new(),
        }
    }

    /// Extracts GraphQL content from a tagged template literal if it matches our criteria.
    ///
    /// This method checks if the tagged template uses `gql` or `graphql` as the tag,
    /// and if so, extracts the template content.
    ///
    /// Skips templates with interpolation since those cannot be statically analyzed
    ///
    fn extract_graphql_from_tagged_template(&mut self, tpl: &TaggedTpl) {
        // Check if this is a GraphQL tagged template (gql`...` or graphql`...`)
        if let Expr::Ident(ident) = &*tpl.tag {
            if ident.sym.as_ref() == "gql" || ident.sym.as_ref() == "graphql" {
                // Skip template literals with interpolation (more than one quasi)
                // These cannot be statically analyzed since the interpolated values
                // are determined at runtime and could affect GraphQL structure
                if tpl.tpl.quasis.len() > 1 {
                    return;
                }

                // Extract the GraphQL content from the template literal
                if let Some(first_quasi) = tpl.tpl.quasis.first() {
                    let content = first_quasi.raw.as_ref();
                    self.graphql_strings.push(GraphQLString {
                        content: content.to_string(),
                        file_path: self.file_path.clone(),
                        position: first_quasi.span.lo().0,
                    });
                }
            }
        }
    }
}

/// Implementation of SWC's Visit trait for GraphQLVisitor.
impl Visit for GraphQLVisitor {
    /// Visits tagged template literal nodes in the AST.
    ///
    /// This method is called automatically by SWC's visitor framework whenever
    /// a tagged template literal is encountered during AST traversal. It processes
    /// the node for GraphQL extraction and then continues traversing child nodes.
    fn visit_tagged_tpl(&mut self, tpl: &TaggedTpl) {
        // Process this tagged template for GraphQL extraction
        self.extract_graphql_from_tagged_template(tpl);

        // Continue visiting child nodes in the AST
        tpl.visit_children_with(self);
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::fs;
    use std::path::PathBuf;

    /// Formats extraction results for snapshot testing with relative file paths.
    ///
    /// This helper function converts extraction results into a consistent string format
    /// suitable for snapshot testing. It uses relative paths from the git root to
    /// ensure snapshots are portable across different development environments.
    ///
    /// # Arguments
    /// * `file_path` - Absolute path to the processed file
    /// * `graphql_strings` - List of extracted GraphQL strings
    ///
    /// # Returns
    /// Formatted string showing file path, count, and each GraphQL string with metadata
    fn format_extraction_result(file_path: &Path, graphql_strings: &[GraphQLString]) -> String {
        // Convert to relative path from git root for portable snapshots
        let git_root = PathBuf::from(env!("CARGO_MANIFEST_DIR"))
            .parent()
            .unwrap()
            .to_path_buf();
        let relative_path = file_path.strip_prefix(&git_root).unwrap_or(file_path);

        let mut result = format!("File: {}\n", relative_path.display());
        result.push_str(&format!(
            "GraphQL strings found: {}\n\n",
            graphql_strings.len()
        ));

        for (i, gql_string) in graphql_strings.iter().enumerate() {
            result.push_str(&format!("=== GraphQL String {} ===\n", i + 1));
            result.push_str(&format!("BytePos: {}\n", gql_string.position));
            result.push_str("Content:\n");
            result.push_str(&gql_string.content);
            result.push_str("\n\n");
        }

        result
    }

    /// Tests extraction from all files in a fixture directory.
    ///
    /// This helper function processes all TypeScript/TSX files in the specified
    /// fixture directory and formats the results for snapshot comparison. It ensures
    /// consistent ordering by sorting files by name.
    ///
    /// # Arguments
    /// * `dir_name` - Name of the fixture directory (valid, invalid, edge_cases)
    ///
    /// # Returns
    /// Formatted string containing extraction results from all files in the directory
    fn test_fixture_directory(dir_name: &str) -> String {
        let fixture_dir = PathBuf::from(env!("CARGO_MANIFEST_DIR"))
            .parent()
            .unwrap()
            .join("fixtures")
            .join(dir_name);

        let mut results = Vec::new();

        if let Ok(entries) = fs::read_dir(&fixture_dir) {
            let mut files: Vec<_> = entries
                .filter_map(|entry| entry.ok())
                .filter(|entry| {
                    let path = entry.path();
                    path.is_file()
                        && (path.extension() == Some(std::ffi::OsStr::new("ts"))
                            || path.extension() == Some(std::ffi::OsStr::new("tsx")))
                })
                .collect();

            // Sort files by name for consistent snapshot ordering
            files.sort_by_key(|entry| entry.file_name());

            for entry in files {
                let file_path = entry.path();
                match extract_graphql_from_file(&file_path) {
                    Ok(graphql_strings) => {
                        let result = format_extraction_result(&file_path, &graphql_strings);
                        results.push(result);
                    }
                    Err(e) => {
                        results.push(format!("File: {}\nError: {}\n\n", file_path.display(), e));
                    }
                }
            }
        }

        results.join("---\n\n")
    }

    /// Tests GraphQL extraction from all valid fixture files.
    ///
    /// This test processes all files in the `fixtures/valid/` directory and creates
    /// a snapshot of the extraction results. Valid fixtures should contain properly
    /// structured GraphQL with appropriate `@catch` and `@throwOnFieldError` directives.
    #[test]
    fn test_valid_fixtures() {
        let result = test_fixture_directory("valid");
        insta::assert_snapshot!(result);
    }

    /// Tests GraphQL extraction from all invalid fixture files.
    ///
    /// This test processes all files in the `fixtures/invalid/` directory and creates
    /// a snapshot of the extraction results. Invalid fixtures contain GraphQL that
    /// violates the `@catch` protection rules.
    #[test]
    fn test_invalid_fixtures() {
        let result = test_fixture_directory("invalid");
        insta::assert_snapshot!(result);
    }

    /// Tests GraphQL extraction from all edge case fixture files.
    ///
    /// This test processes all files in the `fixtures/edge_cases/` directory and creates
    /// a snapshot of the extraction results. Edge cases include scenarios like commented
    /// GraphQL, circular fragment dependencies, and complex nesting patterns.
    #[test]
    fn test_edge_case_fixtures() {
        let result = test_fixture_directory("edge_cases");
        insta::assert_snapshot!(result);
    }
}
