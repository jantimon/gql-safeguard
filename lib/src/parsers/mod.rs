//! Two-stage parsing: extract GraphQL from TypeScript, then parse to AST
//!
//! Separation enables robust extraction from complex TS/TSX without GraphQL syntax errors.

pub mod graphql_parser;
pub mod typescript_parser;
