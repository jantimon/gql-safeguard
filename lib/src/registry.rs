//! GraphQL Extraction Registry
//!

/// A dashmap registry with:
/// process_files()
///
/// a dashmap regisry for all fragments (key fragment name)
/// a dashmap registry for all queries (key query name)
use anyhow::Result;
use dashmap::DashMap;
use globset::{Glob, GlobSetBuilder};
use ignore::{WalkBuilder, WalkState};
use rayon::prelude::*;
use serde::{Deserialize, Serialize};
use std::path::Path;
use std::sync::Arc;

use crate::parsers::graphql_parser::{
    parse_graphql_to_ast, FragmentDefinition, GraphQLItem, QueryOperation,
};
use crate::parsers::typescript_parser::extract_graphql_from_file;

/// Registry for GraphQL fragments
pub type FragmentRegistry = Arc<DashMap<String, FragmentDefinition>>;

/// Registry for GraphQL queries
pub type QueryRegistry = Arc<DashMap<String, QueryOperation>>;

/// Main registry that holds both fragments and queries
#[derive(Serialize, Deserialize)]
pub struct GraphQLRegistry {
    #[serde(with = "serde_dashmap")]
    pub fragments: FragmentRegistry,
    #[serde(with = "serde_dashmap")]
    pub queries: QueryRegistry,
}

/// Custom serialization for DashMap
mod serde_dashmap {
    use super::*;
    use serde::{Deserializer, Serializer};
    use std::collections::HashMap;

    pub fn serialize<S, K, V>(
        dashmap: &Arc<DashMap<K, V>>,
        serializer: S,
    ) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
        K: Clone + Serialize + Eq + std::hash::Hash,
        V: Clone + Serialize,
    {
        let map: HashMap<K, V> = dashmap
            .iter()
            .map(|entry| (entry.key().clone(), entry.value().clone()))
            .collect();
        map.serialize(serializer)
    }

    pub fn deserialize<'de, D, K, V>(deserializer: D) -> Result<Arc<DashMap<K, V>>, D::Error>
    where
        D: Deserializer<'de>,
        K: Clone + Deserialize<'de> + Eq + std::hash::Hash,
        V: Clone + Deserialize<'de>,
    {
        let map: HashMap<K, V> = HashMap::deserialize(deserializer)?;
        let dashmap = DashMap::new();
        for (k, v) in map {
            dashmap.insert(k, v);
        }
        Ok(Arc::new(dashmap))
    }
}

impl Default for GraphQLRegistry {
    fn default() -> Self {
        Self::new()
    }
}

impl GraphQLRegistry {
    pub fn new() -> Self {
        Self {
            fragments: Arc::new(DashMap::new()),
            queries: Arc::new(DashMap::new()),
        }
    }
}

/// Function which returns a new GraphQL registry for the given file list
pub fn process_files(files: &[String]) -> GraphQLRegistry {
    let registry = GraphQLRegistry::new();

    files.par_iter().for_each(|file| {
        parse_file(Path::new(file), &registry);
    });

    registry
}

/// Function which processes files matching a glob pattern using streaming
pub fn process_glob(
    root_path: &Path,
    include_patterns: &[&str], // e.g. &["**/*.ts", "**/*.tsx"]
    exclude_patterns: &[&str], // e.g. &["**/node_modules/**"]
) -> Result<GraphQLRegistry> {
    // Build include GlobSet
    let mut include_builder = GlobSetBuilder::new();
    for pattern in include_patterns {
        include_builder.add(Glob::new(pattern)?);
    }
    let include_set = Arc::new(include_builder.build()?);

    // Build exclude GlobSet
    let mut exclude_builder = GlobSetBuilder::new();
    for pattern in exclude_patterns {
        exclude_builder.add(Glob::new(pattern)?);
    }
    let exclude_set = Arc::new(exclude_builder.build()?);

    let registry = GraphQLRegistry::new();
    let registry_ref = &registry;

    WalkBuilder::new(root_path)
        .standard_filters(false)
        .build_parallel()
        .run(|| {
            let include = Arc::clone(&include_set);
            let exclude = Arc::clone(&exclude_set);
            let registry = registry_ref;

            Box::new(move |entry_res: Result<ignore::DirEntry, ignore::Error>| {
                if let Ok(entry) = entry_res {
                    let path = entry.path();
                    if path.is_dir() && exclude.is_match(path) {
                        return WalkState::Skip;
                    } else if path.is_file() && include.is_match(path) {
                        parse_file(path, registry);
                    }
                }
                WalkState::Continue
            })
        });

    Ok(registry)
}

fn parse_file(file: &Path, registry: &GraphQLRegistry) {
    let graphql_strings_result = extract_graphql_from_file(Path::new(file));
    if let Ok(graphql_strings) = graphql_strings_result {
        for graphql_string in &graphql_strings {
            let graphql_ast = parse_graphql_to_ast(graphql_string);
            if let Ok(ast) = graphql_ast {
                for graphql_item in ast {
                    match graphql_item {
                        GraphQLItem::Fragment(fragment) => {
                            registry.fragments.insert(fragment.name.clone(), fragment);
                        }
                        GraphQLItem::Query(query) => {
                            registry.queries.insert(query.name.clone(), query);
                        }
                    }
                }
            }
        }
    } else {
        eprintln!("Failed to parse GraphQL from file: {}", file.display());
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::tree_formatter::TreeFormatter;
    use std::fs;
    use std::path::PathBuf;

    /// Formats a GraphQL registry using TreeFormatter for snapshot testing
    fn format_registry_with_tree_formatter(registry: &GraphQLRegistry) -> String {
        let mut formatter = TreeFormatter::new();

        // Git root for relative paths in snapshots
        let git_root = PathBuf::from(env!("CARGO_MANIFEST_DIR"))
            .parent()
            .unwrap()
            .to_path_buf();

        formatter.add_line(0, "GraphQL Registry");

        // Format queries section
        let query_count = registry.queries.len();
        formatter.add_line(1, &format!("Queries ({})", query_count));

        if query_count > 0 {
            // Sort queries by name for consistent output
            let mut queries: Vec<_> = registry.queries.iter().collect();
            queries.sort_by(|a, b| a.key().cmp(b.key()));

            for entry in queries {
                let name = entry.key();
                let query = entry.value();
                let relative_path = query
                    .file_path
                    .strip_prefix(&git_root)
                    .unwrap_or(&query.file_path);
                formatter.add_line(2, &format!("{} ({})", name, relative_path.display()));

                // Query directives
                if !query.directives.is_empty() {
                    formatter.add_line(3, "Directives:");
                    for directive in &query.directives {
                        let emoji = match directive.directive_type {
                            crate::parsers::graphql_parser::DirectiveType::Catch => "üß§",
                            crate::parsers::graphql_parser::DirectiveType::ThrowOnFieldError => {
                                "‚ö†Ô∏è"
                            }
                        };
                        formatter.add_line(4, &format!("{:?} {}", directive.directive_type, emoji));
                    }
                }

                // Query fields
                let query_fields = query.fields();
                if !query_fields.is_empty() {
                    formatter.add_line(3, "Fields:");
                    for field in &query_fields {
                        let mut field_text = field.name.clone();
                        if !field.directives.is_empty() {
                            let directive_strs: Vec<String> = field.directives.iter().map(|d| {
                                let emoji = match d.directive_type {
                                    crate::parsers::graphql_parser::DirectiveType::Catch => "üß§",
                                    crate::parsers::graphql_parser::DirectiveType::ThrowOnFieldError => "‚ö†Ô∏è",
                                };
                                format!("{:?} {}", d.directive_type, emoji)
                            }).collect();
                            field_text.push_str(&format!(" [{}]", directive_strs.join(", ")));
                        }
                        formatter.add_line(4, &field_text);
                    }
                }

                // Query fragment spreads
                let query_fragments = query.fragments();
                if !query_fragments.is_empty() {
                    formatter.add_line(3, "Fragment Spreads:");
                    for fragment in &query_fragments {
                        let mut fragment_text = fragment.name.clone();
                        if !fragment.directives.is_empty() {
                            let directive_strs: Vec<String> = fragment.directives.iter().map(|d| {
                                let emoji = match d.directive_type {
                                    crate::parsers::graphql_parser::DirectiveType::Catch => "üß§",
                                    crate::parsers::graphql_parser::DirectiveType::ThrowOnFieldError => "‚ö†Ô∏è",
                                };
                                format!("{:?} {}", d.directive_type, emoji)
                            }).collect();
                            fragment_text.push_str(&format!(" [{}]", directive_strs.join(", ")));
                        }
                        formatter.add_line(4, &fragment_text);
                    }
                }
            }
        }

        // Format fragments section
        let fragment_count = registry.fragments.len();
        formatter.add_line(1, &format!("Fragments ({})", fragment_count));

        if fragment_count > 0 {
            // Sort fragments by name for consistent output
            let mut fragments: Vec<_> = registry.fragments.iter().collect();
            fragments.sort_by(|a, b| a.key().cmp(b.key()));

            for entry in fragments {
                let name = entry.key();
                let fragment = entry.value();
                let relative_path = fragment
                    .file_path
                    .strip_prefix(&git_root)
                    .unwrap_or(&fragment.file_path);
                formatter.add_line(2, &format!("{} ({})", name, relative_path.display()));

                // Fragment directives
                if !fragment.directives.is_empty() {
                    formatter.add_line(3, "Directives:");
                    for directive in &fragment.directives {
                        let emoji = match directive.directive_type {
                            crate::parsers::graphql_parser::DirectiveType::Catch => "üß§",
                            crate::parsers::graphql_parser::DirectiveType::ThrowOnFieldError => {
                                "‚ö†Ô∏è"
                            }
                        };
                        formatter.add_line(4, &format!("{:?} {}", directive.directive_type, emoji));
                    }
                }

                // Fragment fields
                let fragment_fields = fragment.fields();
                if !fragment_fields.is_empty() {
                    formatter.add_line(3, "Fields:");
                    for field in &fragment_fields {
                        let mut field_text = field.name.clone();
                        if !field.directives.is_empty() {
                            let directive_strs: Vec<String> = field.directives.iter().map(|d| {
                                let emoji = match d.directive_type {
                                    crate::parsers::graphql_parser::DirectiveType::Catch => "üß§",
                                    crate::parsers::graphql_parser::DirectiveType::ThrowOnFieldError => "‚ö†Ô∏è",
                                };
                                format!("{:?} {}", d.directive_type, emoji)
                            }).collect();
                            field_text.push_str(&format!(" [{}]", directive_strs.join(", ")));
                        }
                        formatter.add_line(4, &field_text);
                    }
                }

                // Fragment spreads
                let fragment_spreads = fragment.fragments();
                if !fragment_spreads.is_empty() {
                    formatter.add_line(3, "Fragment Spreads:");
                    for spread in &fragment_spreads {
                        let mut spread_text = spread.name.clone();
                        if !spread.directives.is_empty() {
                            let directive_strs: Vec<String> = spread.directives.iter().map(|d| {
                                let emoji = match d.directive_type {
                                    crate::parsers::graphql_parser::DirectiveType::Catch => "üß§",
                                    crate::parsers::graphql_parser::DirectiveType::ThrowOnFieldError => "‚ö†Ô∏è",
                                };
                                format!("{:?} {}", d.directive_type, emoji)
                            }).collect();
                            spread_text.push_str(&format!(" [{}]", directive_strs.join(", ")));
                        }
                        formatter.add_line(4, &spread_text);
                    }
                }
            }
        }

        formatter.to_string()
    }

    /// Collects all TypeScript/TSX files from a fixture directory
    fn collect_fixture_files(dir_name: &str) -> Vec<String> {
        let fixture_dir = PathBuf::from(env!("CARGO_MANIFEST_DIR"))
            .parent()
            .unwrap()
            .join("fixtures")
            .join(dir_name);

        let mut files = Vec::new();

        if let Ok(entries) = fs::read_dir(&fixture_dir) {
            let mut file_entries: Vec<_> = entries
                .filter_map(|entry| entry.ok())
                .filter(|entry| {
                    let path = entry.path();
                    path.is_file()
                        && (path.extension() == Some(std::ffi::OsStr::new("ts"))
                            || path.extension() == Some(std::ffi::OsStr::new("tsx")))
                })
                .collect();

            // Sort files by name for consistent ordering
            file_entries.sort_by_key(|entry| entry.file_name());

            for entry in file_entries {
                files.push(entry.path().to_string_lossy().to_string());
            }
        }

        files
    }

    /// Tests registry building from valid fixture files
    #[test]
    fn test_registry_from_valid_fixtures() {
        let files = collect_fixture_files("valid");
        let registry = process_files(&files);
        let formatted = format_registry_with_tree_formatter(&registry);
        insta::assert_snapshot!(formatted);
    }

    /// Tests registry building from invalid fixture files
    #[test]
    fn test_registry_from_invalid_fixtures() {
        let files = collect_fixture_files("invalid");
        let registry = process_files(&files);
        let formatted = format_registry_with_tree_formatter(&registry);
        insta::assert_snapshot!(formatted);
    }

    /// Tests registry building from edge case fixture files
    #[test]
    fn test_registry_from_edge_case_fixtures() {
        let files = collect_fixture_files("edge_cases");
        let registry = process_files(&files);
        let formatted = format_registry_with_tree_formatter(&registry);
        insta::assert_snapshot!(formatted);
    }

    /// Tests registry building from all fixture files combined
    #[test]
    fn test_registry_from_all_fixtures() {
        let mut all_files = Vec::new();
        all_files.extend(collect_fixture_files("valid"));
        all_files.extend(collect_fixture_files("invalid"));
        all_files.extend(collect_fixture_files("edge_cases"));

        let registry = process_files(&all_files);
        let formatted = format_registry_with_tree_formatter(&registry);
        insta::assert_snapshot!(formatted);
    }
}
