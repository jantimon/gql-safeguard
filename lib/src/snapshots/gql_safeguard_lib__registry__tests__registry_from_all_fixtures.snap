---
source: lib/src/registry.rs
expression: formatted
---
GraphQL Registry
├── Queries (16)
|   ├── AdditionalQueryEdgeCase (fixtures/edge_cases/additional-queries.ts)
|   |   └── Fields:
|   |       ├── user
|   |       ├── id
|   |       └── importedField [ThrowOnFieldError ⚠️]
|   ├── BaseQueryDynamic (fixtures/edge_cases/dynamic_imports.tsx)
|   |   ├── Directives:
|   |   |   └── Catch 🧤
|   |   └── Fields:
|   |       ├── user
|   |       ├── id
|   |       ├── name
|   |       └── dynamicField [ThrowOnFieldError ⚠️]
|   ├── CircularQueryTest (fixtures/edge_cases/circular_fragments.ts)
|   |   ├── Fields:
|   |   |   └── user
|   |   └── Fragment Spreads:
|   |       └── FragmentACircular
|   ├── GetFullUser (fixtures/valid/nested_fragments_protected.ts)
|   |   ├── Fields:
|   |   |   └── user
|   |   └── Fragment Spreads:
|   |       └── UserDetails
|   ├── GetFullUserUnprotected (fixtures/invalid/unprotected_nested.ts)
|   |   ├── Fields:
|   |   |   └── user
|   |   └── Fragment Spreads:
|   |       └── UserDetailsUnprotected
|   ├── GetUserBasic (fixtures/valid/required_throw.ts)
|   |   ├── Directives:
|   |   |   └── Catch 🧤
|   |   └── Fields:
|   |       ├── user
|   |       ├── id
|   |       ├── name [RequiredThrow ⚠️]
|   |       └── email [RequiredThrow ⚠️]
|   ├── GetUserMixed (fixtures/valid/required_throw.ts)
|   |   ├── Directives:
|   |   |   └── Catch 🧤
|   |   └── Fields:
|   |       ├── user
|   |       ├── id
|   |       ├── name [RequiredThrow ⚠️]
|   |       ├── avatar [ThrowOnFieldError ⚠️]
|   |       └── email [RequiredThrow ⚠️]
|   ├── GetUserNested (fixtures/valid/required_throw.ts)
|   |   └── Fields:
|   |       ├── user
|   |       ├── id
|   |       ├── profile [Catch 🧤]
|   |       ├── displayName [RequiredThrow ⚠️]
|   |       ├── bio [RequiredThrow ⚠️]
|   |       ├── avatar
|   |       └── url [ThrowOnFieldError ⚠️]
|   ├── GetUserProfile (fixtures/valid/required_throw.ts)
|   |   └── Fields:
|   |       ├── user [Catch 🧤]
|   |       ├── id
|   |       ├── name [RequiredThrow ⚠️]
|   |       ├── avatar [RequiredThrow ⚠️]
|   |       └── bio
|   ├── GetUserProfileUnprotected (fixtures/invalid/missing_catch.tsx)
|   |   └── Fields:
|   |       ├── user
|   |       ├── id
|   |       ├── name
|   |       ├── avatar [ThrowOnFieldError ⚠️]
|   |       └── email
|   ├── GetUserUnprotected (fixtures/invalid/unprotected_required_throw.ts)
|   |   └── Fields:
|   |       ├── user
|   |       ├── id
|   |       ├── name [RequiredThrow ⚠️]
|   |       └── email
|   ├── GetUserWithRiskyData (fixtures/valid/fragment_level_catch.ts)
|   |   ├── Directives:
|   |   |   └── Catch 🧤
|   |   ├── Fields:
|   |   |   ├── user
|   |   |   ├── id
|   |   |   ├── name
|   |   |   └── friends [ThrowOnFieldError ⚠️]
|   |   └── Fragment Spreads:
|   |       └── RiskyUserData
|   ├── GraphQLQuery (fixtures/edge_cases/additional-queries.ts)
|   |   ├── Directives:
|   |   |   └── Catch 🧤
|   |   └── Fields:
|   |       ├── user
|   |       ├── id
|   |       └── protectedField [ThrowOnFieldError ⚠️]
|   ├── MixedQuery (fixtures/invalid/partial_protection.ts)
|   |   ├── Fields:
|   |   |   ├── user
|   |   |   └── id
|   |   └── Fragment Spreads:
|   |       ├── ProtectedFragment
|   |       └── UnprotectedFragment
|   ├── StaticDynamicQueryEdge (fixtures/edge_cases/dynamic_imports.tsx)
|   |   └── Fields:
|   |       ├── user
|   |       ├── id
|   |       └── dynamicField [ThrowOnFieldError ⚠️]
|   └── ValidQueryCommented (fixtures/edge_cases/commented_graphql.ts)
|       ├── Directives:
|       |   └── Catch 🧤
|       └── Fields:
|           ├── user
|           ├── id
|           ├── name
|           └── avatar [ThrowOnFieldError ⚠️]
└── Fragments (13)
    ├── FragmentACircular (fixtures/edge_cases/circular_fragments.ts)
    |   ├── Fields:
    |   |   ├── id
    |   |   └── name
    |   └── Fragment Spreads:
    |       └── FragmentBCircular
    ├── FragmentBCircular (fixtures/edge_cases/circular_fragments.ts)
    |   ├── Fields:
    |   |   ├── email
    |   |   └── bio
    |   └── Fragment Spreads:
    |       └── FragmentACircular
    ├── ProtectedFragment (fixtures/invalid/partial_protection.ts)
    |   ├── Directives:
    |   |   └── Catch 🧤
    |   └── Fields:
    |       ├── sensitiveData [ThrowOnFieldError ⚠️]
    |       └── otherData
    ├── RiskyUserData (fixtures/valid/fragment_level_catch.ts)
    |   ├── Directives:
    |   |   └── Catch 🧤
    |   └── Fields:
    |       ├── sensitiveInfo [ThrowOnFieldError ⚠️]
    |       ├── backupEmail [ThrowOnFieldError ⚠️]
    |       └── internalId
    ├── UnprotectedFragment (fixtures/invalid/partial_protection.ts)
    |   └── Fields:
    |       ├── riskyField [ThrowOnFieldError ⚠️]
    |       └── normalField
    ├── UserAvatar (fixtures/valid/nested_fragments_protected.ts)
    |   ├── Directives:
    |   |   └── Catch 🧤
    |   ├── Fields:
    |   |   └── avatar
    |   └── Fragment Spreads:
    |       └── avatarImage
    ├── UserAvatarUnprotected (fixtures/invalid/unprotected_nested.ts)
    |   ├── Directives:
    |   |   └── ThrowOnFieldError ⚠️
    |   └── Fields:
    |       ├── avatar
    |       └── avatarUrl
    ├── UserBasicInfo (fixtures/valid/nested_fragments_protected.ts)
    |   └── Fields:
    |       ├── id
    |       ├── name
    |       └── email
    ├── UserBasicInfoUnprotected (fixtures/invalid/unprotected_nested.ts)
    |   └── Fields:
    |       ├── id
    |       ├── name [ThrowOnFieldError ⚠️]
    |       └── email
    ├── UserBio (fixtures/valid/nested_fragments_protected.ts)
    |   └── Fields:
    |       ├── bioText
    |       └── bioImage [ThrowOnFieldError ⚠️]
    ├── UserDetails (fixtures/valid/nested_fragments_protected.ts)
    |   ├── Fields:
    |   |   ├── details [Catch 🧤]
    |   |   └── bio [ThrowOnFieldError ⚠️]
    |   └── Fragment Spreads:
    |       ├── UserBasicInfo
    |       ├── UserAvatar
    |       └── UserBio
    ├── UserDetailsUnprotected (fixtures/invalid/unprotected_nested.ts)
    |   ├── Fields:
    |   |   └── bio
    |   └── Fragment Spreads:
    |       ├── UserBasicInfoUnprotected
    |       └── UserAvatarUnprotected
    └── avatarImage (fixtures/valid/nested_fragments_protected.ts)
        ├── Directives:
        |   └── ThrowOnFieldError ⚠️
        └── Fields:
            ├── avatar
            └── avatarUrl
